### PR TITLE
Ajax errors

### PR DESCRIPTION
**Pull request by curiousrohan1**

### CODE REVIEW

File: src/main/java/org/dishorg/dishorg/RestExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+@ControllerAdvice
+public class RestExceptionHandler extends ResponseEntityExceptionHandler {
+    @ExceptionHandler(value = {DuplicateRecipeException.class})
+    protected ResponseEntity<Object> handleConflict(RuntimeException ex, WebRequest request) {
+        return handleExceptionInternal(ex, new ApiError(HttpStatus.BAD_REQUEST, ex.getMessage()),
+                new HttpHeaders(), HttpStatus.BAD_REQUEST, request);
+    }
+
+    private static class ApiError {
```
Comment (stamhankar999) [inline src/main/java/org/dishorg/dishorg/RestExceptionHandler.java:21]: Add an `@Getter` Lombok annotation here, and then we can remove the explicit getter methods.

File: src/main/java/org/dishorg/dishorg/RestExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        private final int statusCode;
+        private final HttpStatus status;
+        private final String message;
+
+        public ApiError(HttpStatus status, String message) {
+            super();
```
Comment (stamhankar999) [inline src/main/java/org/dishorg/dishorg/RestExceptionHandler.java]: Remove this call to `super()`. It's unnecessary and happens automatically when constructing ApiError.

File: src/main/java/org/dishorg/dishorg/RestExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            this.statusCode = status.value();
+            this.status = status;
+            this.message = message;
+        }
+
+        public int getStatusCode() {
```
Comment (stamhankar999) [inline src/main/java/org/dishorg/dishorg/RestExceptionHandler.java]: Remove these getters.

File: src/main/resources/public/js/recipes.js [hunk 4]
```diff
@@ -80,11 +80,13 @@ function ingApply(idx) {
...
     contentType: 'application/json',
     dataType: 'json',
-  }).done((data) => {
-    successOnAjaxOfRecipe(data);
-  }).fail((jqXHR, textStatus) => {
-    alert(`${textStatus}, ${jqXHR}`);
-  });
+  }).fail(
+    (jqXHR) => {
+      $('#error-message-dialogue').text(jqXHR.responseJSON.message);
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js]: I see that all of your fail handlers are doing the same thing; a long time ago you asked about making a common function to have that logic. I think that time has come. Please refactor to a function that takes a jqXHR and update all the fail handlers accordingly.

File: src/main/resources/public/js/recipes.js [hunk 5]
```diff
@@ -111,18 +114,18 @@ function successOnAjaxOfRecipe(recipe) {
...
-            <button class="edit-recipes btn" id="edit-ing${idx}" data-idx="${idx}">
-                <img src="images/edit.jpg"style="width:30px;height:30px;">
-            </button>
-            <button type="button" class="btn openModal" data-toggle="modal" data-idx="${idx}"
-                    id="open-modal${idx}" data-target="#cont-del-modal">
-                <img src="images/del.png"style="width:30px;height:30px;">
-            </button>
-            ${line}
-        </li>`);
+      <li id="ing-line-container${idx}">
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js:120]: I'm guessing the code formatter did this. I don't know how to keep it from reformatting this, but fine...

File: src/main/resources/public/js/recipes.js [hunk 6]
```diff
@@ -259,23 +274,38 @@ $(document).ready(() => {
...
       dataType: 'json',
-      success: (recipe) => {
+    }).done(
+      (recipe) => {
+        currentRec = recipe;
         $('#recipe-list').append(
           `<button class="recipe-list list-group-item list-group-item-action btn" type="button"
-                   data-id="${recipe.id}">${recipe.name}
-           </button>`,
+                                                            data-id="${recipe.id}">${recipe.name}
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js:304]: This is really weird formatting. Was this from the code formatter?

File: src/main/resources/public/js/recipes.js [hunk 7]
```diff
@@ -259,23 +274,38 @@ $(document).ready(() => {
...
         lastButton.trigger('click');
+        $('#plus-ing').click();
       },
-    });
+    ).fail(
+      (jqXHR) => {
+        $('#add-rec-div').show();
+        $('#error-message-dialogue').text(jqXHR.responseJSON.message);
+        $('#error-message').show();
+        //         jqXHRg=jqXHR;
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js]: Remove these commented-out lines.

File: src/main/resources/public/js/recipes.js [hunk 8]
```diff
@@ -259,23 +274,38 @@ $(document).ready(() => {
...
         );
         const lastButton = $('button.recipe-list.list-group-item-action:last-child');
         addRecHandler(lastButton);
         lastButton.trigger('click');
+        $('#plus-ing').click();
       },
-    });
+    ).fail(
+      (jqXHR) => {
+        $('#add-rec-div').show();
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js:314]: Oh, this is a little different from the other fail handlers. So for this one, keep this, but call the common function to do the error message rendering that is common for all fail handlers.

File: src/main/resources/public/js/recipes.js [hunk 9]
```diff
@@ -288,10 +318,16 @@ $(document).ready(() => {
...
       url: `/recipes/${currentRec.id}`,
       dataType: 'json',
-      statusCode: {
-        200: reset,
-      },
-    });
+    })
+      .fail(
+        (jqXHR, status, errorThrown) => {
+          alert('jqXHR:');
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js]: Shouldn't this be doing the same failure handling as the other places?

File: src/main/resources/public/js/recipes.js [hunk 10]
```diff
@@ -301,16 +337,34 @@ $(document).ready(() => {
...
+    })
+      .done(
+        (data) => {
+          $('button.recipe-list.active').text(data.name);
+        },
+      )
+      .fail(
+        (jqXHR) => {
+          $('#renamed-recipe-name').show();
+          $('#error-message-dialogue').text(jqXHR.responseJSON.message);
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js]: Same here; do logic specific to this scenario here, then call the common function.

File: src/main/resources/public/js/recipes.js [hunk 11]
```diff
@@ -301,16 +337,34 @@ $(document).ready(() => {
...
+          $('#error-message').show();
+          // alert('jqXHR:');
```
Comment (stamhankar999) [inline src/main/resources/public/js/recipes.js]: Get rid of commented out code.

File: src/main/resources/public/recipes.html [hunk 12]
```diff
@@ -57,44 +57,52 @@ <h5 class="modal-title" id="cont-del-ing">Are you sure?</h5>
...
-    <h3><strong>Recipes:</strong>
-        <button class="btn" data-placement="right" data-toggle="tooltip" id="plus-rec"
-                style="float:right;width:3rem;height:3rem;"
-                title="Add recipe">+
-        </button>
-    </h3>
-    <hr>
-<!--    <div data-offset="0" data-spy="scroll" data-target="#recipe-list">-->
+<div class="alert-danger" id="error-message" style="display:none;text-align:center;">
+    <h3>Error</h3>
```
Comment (stamhankar999) [inline src/main/resources/public/recipes.html]: When looking at this running on our Raspberry PI, I feel like this "Error" header is too much. Why not just have the error message?

**Review by stamhankar999 — CHANGES_REQUESTED** [review]
Looks good overall. I have various comments I'd like you to address.

Also, I am running your branch in our Raspberry PI, and I noticed an issue where once an error appears, it doesn't go away when I do other actions. For example, after trying to add a duplicate recipe and failing, I click on a different recipe to load it, and the error message remains on-screen. The error should be hidden before doing any new action.

File: src/main/resources/public/js/recipes.js (hunk 6 above)
Comment (curiousrohan1) [reply src/main/resources/public/js/recipes.js:304]: Yeah, it is. Should I fix that?

Comment (stamhankar999) [reply src/main/resources/public/js/recipes.js:304]: Well, if the code formatter is going to change it back, then don't bother.

Comment (stamhankar999) [reply src/main/resources/public/js/recipes.js:304]: But if it'll stick to where it should be, then yeah, fix it.

**Review by stamhankar999 — APPROVED** [review]

**Pull request merged**